redirects-list.json

```diff
@@ -1,8 +1,2 @@
 [
-  "^https://(?:www\\.)?cml\\.dev(.*)?                                      https://cml.dev$1",
```

Comment:
It feels we should keep some of those:

www - for sure,
docs/documentation - also
the last one that removes the `/` also
everything else - cc @casperdcl - do you know why do we have them?

Reply:
@shcheklein  I think you're right. that's why I never removed that line before. however, in this pr, i am trying to remove all redirections to test. the issue that you mentioned on feedback is really odd for me.
did you have a chance to check the live server's configuration?

Reply:
> did you have a chance to check the live server's configuration?

@charnysh-dev could you clarify please, what do you mean?

> in this pr, i am trying to remove all redirections to test.

we can deploy it to test it on prod, I don't see any problem with that. Let me know.

Reply:
> > did you have a chance to check the live server's configuration?
> 
> @charnysh-dev could you clarify please, what do you mean?
> 
@shcheklein i know the odd redirection issues never happened on the prev test (https://github.com/iterative/cml-website/pull/60) as well as the current pr's deployment (https://cml-dev-fix-misc-p5egcxviplyb6.herokuapp.com)
but we see the issues on `cml.dev`

Reply:
@charnysh-dev I see.

Rules look the same to me on the Cloudflare. Heroku doesn't do anything specific with redirects. 

I see that I can reproduce some problems on the https://cml-dev-fix-misc-p5egcxviplyb6.herokuapp.com though.

Open this link, refresh the page. I see that `/` at the end "blinks", also it goes into `http:\\` (not secure).

Reply:
@shcheklein
I have checked & compared the smart link component of CML again to the one in the DVC. But still I haven't found anything else about the strange problems.
Btw, I tried with multiple platforms/devices even, but couldn't reproduce the above problem on the Heroku you mentioned.
Could you please share more details about the Heroku issue?

Reply:
Sure. I'm able to reproduce it if open directly https://cml-dev-fix-misc-p5egcxviplyb6.herokuapp.com/doc

That's what I see after loading is done:

<img width="1439" alt="Screen Shot 2021-07-18 at 11 01 54 PM" src="https://user-images.githubusercontent.com/3659196/126110413-7e87d6a1-ae9c-4ac3-b840-5a7e5ad35de6.png">

It's not secure for some reason - changes https to http, but this might be Heroku preview env specific, not sure. Also you can see that it redirects to `doc/` (and then `/` is dropped in JS?). 

@rogermparent you might remember some specific logic we applied at the end to Gatsby + express (?) to get rid of `/` and there redirects on `dvc.org`?

Reply:
@shcheklein @charnysh-dev I can't definitively know if any logic in the code makes or breaks the redirects feature here. That said, I do have a few comments that may help. I'm going under the assumption that URL rewrites on the server side will be necessary, but it's possible that I'm mistaken since base cml.dev doesn't have the slash-appending issue described here unlike the docs pages.

> Gatsby + express (?)

I'm pretty sure cml.dev still doesn't run express even though the server code is in there from the original docs engine port, and that's due to `heroku-buildpack-static` which is the normal way to run static site generators like Gatsby on Heroku by provisioning a standard Apache webserver to serve the built site since most SSG repos don't define their own servers.

You can see in the configs of the productions apps that dvc.org doesn't use `heroku-buildpack-static` while cml.dev does.

dvc.org
![Screenshot 2021-07-20 at 20-11-13 dvc-org · Settings Heroku](https://user-images.githubusercontent.com/9111807/126411211-f4358a08-c283-47f9-9a68-04169aaa0b12.png)

cml.dev
![Screenshot 2021-07-20 at 20-09-48 cml-dev · Settings Heroku](https://user-images.githubusercontent.com/9111807/126411221-1876fba3-9be1-44bd-963c-65a2a4f9d0a4.png)

I'm not sure if just removing the buildpack will be all that is necessary to switch cml.dev to express, but I do believe it to be a necessary step and good place to start alongside making the heroku config of cml.dev match dvc.org.

Reply:
Coming back to the fact that the homepage doesn't suffer from this trailing slash issue, there may be some logic in the docs page specifically that's adding the slash before client side code takes it away- I don't know specifically what, especially in the port, but it could be a case of just removing some rogue logic. 

The pages don't have trailing slashes in the underlying Gatsby GraphQL nodes, so we can eliminate that.

Reply:
I tested this against a blank page on CML, as well as a completely new Gatsby site, and found the same issue persists.

After some more digging and reading [this top article](https://ahrefs.com/blog/trailing-slash/), it seems trailing slashes are a part of standard web servers in general: by convention, directories (with clean URLs that means all pages) normally have a trailing slash and as such most regular web servers like nginx and whatever Gatsby uses for its dev server adds them without some extra config.

Seems like we do need server rewrites to enforce having no trailing slashes, which can either be done by reintroducing the Express server and configuring that (which I suspect will happen eventually anyway) or by using CloudFlare-based rewrites as a simpler solution that I'd personally prefer.

Reply:
> CloudFlare-based rewrites as a simpler solution that I'd personally prefer.

As far as I understand regular Gatsby code would add `/` back. The problem is not that we have an URL like `something/`, the problem is that we have `something` and Gatsby adds the `/`. So, I doubt that CF could solve this, but I'm not 100% sure.

Reply:
I really don't think it's Gatsby, there's nothing in Gatsby that's adding the trailing slash client-side as far as I can tell and the behavior I described looks consistent among static sites. As I understand it, the issue goes like this:

1. User requests `cml.dev/doc`
2. nginx from `heroku-buildpack-static` responds with `cml.dev/doc/` because it's a directory and trailing slashes are added by default
3. Gatsby changes the URL to `cml.dev/doc` on the client side during hydration

Since CloudFlare is acting as DNS and a middleman to all requests, its URL rewrites should also have the ability to change the response- I could be wrong, as CloudFlare can be kind of a weird beast in these stacks.

Reply:
Thanks @rogermparent ! @charnysh-dev , let's try to get rid of Gastby static build pack and use exactly the same stack as we do for dvc.org? WDYT?

Reply:
Thank you @rogermparent @shcheklein 
Very helpful for me. let me check check & fix them.

Reply:
I think even if we solve this trailing slash issue through CloudFlare, we're going to want to add the proxy server for one reason or another in the future- might as well be now.

Good luck @charnysh-dev!

Reply:
@shcheklein @rogermparent thanks guys, i cannot find the strange issue after removing the heroku-buildpack-static as you mentioned.
https://cml-dev-fix-misc-ztn7ocuwjc99b.herokuapp.com/doc